test/util/node-mocks-http.js

```diff
@@ -0,0 +1,24 @@
+// Safe wrapper for node-mocks-http
+// See: https://github.com/eugef/node-mocks-http/issues/299
+
+const wrapped = require('node-mocks-http');
+// qs is an implicit dependency of express:
+const qs = require('qs'); // eslint-disable-line import/no-extraneous-dependencies
```

Comment:
> maybe it starts becoming necessary to pass certain options to `qs.parse()` to match what Express does

I don't think express v4 will ever change (1) that it uses `qs` by default, and (2) how it initialises `qs`.  (see e.g. https://github.com/expressjs/express/issues/5723)

> I think it'd be a good idea for us to pin qs. What do you think?

I think that might have the opposite effect - possibly introducing a duplicate `qs` module.  E.g.:

<details>
  <summary>current duplicates</summary>

## `report-dupes.js`

```js
const seen = {};

const { packages } = require('./package-lock');

for(const [ fullName, { version } ] of Object.entries(packages)) {
  const name = fullName.replace(/.*\/node_modules\//, '');
  if(!seen[name]) seen[name] = new Set();

  const ref = name === fullName ?
      '/' :
      fullName
          .replace(/^node_modules\//, '') 
          .replace(new RegExp(`/node_modules/${name}$`), '');
  seen[name].add(`${version} (${ref})`);
}

Object.entries(seen)
    .filter(([ name, versions ]) => versions.size > 1)
    .sort(([a], [b]) => a < b ? -1 : 1)
    .forEach(([ name, versions ]) => console.log(name, [...versions]));

```

## Duplicates

```sh
$ node report-dupes.js 
agent-base [
  '7.1.0 (http-proxy-agent)',
  '7.1.0 (pac-proxy-agent)',
  '7.1.0 (proxy-agent)',
  '7.1.0 (socks-proxy-agent)'
]
ansi-styles [ '3.2.1 (@babel/code-frame)', '3.2.1 (@babel/highlight)' ]
async [
  '2.6.4 (@pm2/io)',
  '2.6.4 (@pm2/js-api)',
  '3.2.3 (prompt)',
  '2.6.4 (vizion)'
]
brace-expansion [ '2.0.1 (mocha/node_modules/minimatch)', '2.0.1 (readdir-glob)' ]
chalk [
  '2.4.2 (@babel/code-frame)',
  '2.4.2 (@babel/highlight)',
  '3.0.0 (@pm2/agent)',
  '3.0.0 (cli-tableau)',
  '3.0.0 (pm2)'
]
color-convert [ '1.9.3 (@babel/code-frame)', '1.9.3 (@babel/highlight)' ]
color-name [ '1.1.3 (@babel/code-frame)', '1.1.3 (@babel/highlight)' ]
commander [ '6.2.1 (knex)', '2.15.1 (pm2)' ]
cookie [ '0.4.1 (cookie-parser)', '0.6.0 (express)' ]
debug [
  '4.3.4 (@babel/core)',
  '4.3.4 (@babel/traverse)',
  '4.3.4 (@eslint/eslintrc)',
  '4.3.4 (@humanwhocodes/config-array)',
  '4.3.4 (@koa/router)',
  '4.3.4 (@pm2/agent)',
  '4.3.7 (@pm2/io)',
  '4.3.4 (@pm2/js-api)',
  '4.3.4 (@pm2/pm2-version-check)',
  '4.3.4 (agent-base)',
  '3.2.7 (eslint-import-resolver-node)',
  '3.2.7 (eslint-module-utils)',
  '4.3.4 (eslint)',
  '4.3.4 (get-uri)',
  '4.3.4 (http-proxy-agent)',
  '4.3.4 (https-proxy-agent)',
  '4.3.4 (istanbul-lib-source-maps)',
  '4.3.1 (knex)',
  '4.3.4 (koa)',
  '4.3.4 (mocha-junit-reporter)',
  '4.3.4 (mocha)',
  '3.2.7 (needle)',
  '4.3.4 (nock)',
  '3.2.7 (nodemon)',
  '4.3.4 (oidc-provider)',
  '4.3.4 (pac-proxy-agent)',
  '4.3.4 (pm2-axon-rpc)',
  '4.3.4 (pm2-axon)',
  '4.3.4 (pm2-sysmonit)',
  '4.3.4 (pm2)',
  '4.3.4 (proxy-agent)',
  '4.3.7 (require-in-the-middle)',
  '4.3.4 (socks-proxy-agent)',
  '4.3.4 (superagent)'
]
define-property [
  '1.0.0 (base)',
  '0.2.5 (class-utils)',
  '0.2.5 (expand-brackets)',
  '1.0.0 (extglob)',
  '0.2.5 (object-copy)',
  '1.0.0 (snapdragon-node)',
  '0.2.5 (snapdragon)',
  '0.2.5 (static-extend)'
]
depd [
  '2.0.0 (body-parser)',
  '2.0.0 (cookies)',
  '2.0.0 (express)',
  '2.0.0 (http-errors)',
  '2.0.0 (koa)',
  '1.1.2 (koa/node_modules/http-errors)',
  '2.0.0 (send)'
]
encodeurl [ '2.0.0 (express)', '2.0.0 (finalhandler)', '2.0.0 (serve-static)' ]
escape-string-regexp [ '1.0.5 (@babel/code-frame)', '1.0.5 (@babel/highlight)' ]
eventemitter2 [ '6.4.9 (@pm2/io)', '6.4.9 (@pm2/js-api)', '0.4.14 (nssocket)' ]
extend-shallow [
  '2.0.1 (expand-brackets)',
  '2.0.1 (extglob)',
  '2.0.1 (findup-sync/node_modules/braces)',
  '2.0.1 (findup-sync/node_modules/fill-range)',
  '2.0.1 (set-value)',
  '2.0.1 (snapdragon)'
]
find-up [
  '4.1.0 (@istanbuljs/load-nyc-config)',
  '4.1.0 (nyc)',
  '4.1.0 (pkg-dir)'
]
has-flag [
  '3.0.0 (@babel/code-frame)',
  '3.0.0 (@babel/highlight)',
  '3.0.0 (nodemon)'
]
http-errors [ '1.8.1 (http-assert)', '1.8.1 (koa)' ]
https-proxy-agent [ '7.0.1 (pac-proxy-agent)', '7.0.1 (proxy-agent)' ]
is-accessor-descriptor [
  '0.1.6 (class-utils)',
  '0.1.6 (expand-brackets)',
  '0.1.6 (object-copy)',
  '0.1.6 (snapdragon)',
  '0.1.6 (static-extend)'
]
is-data-descriptor [
  '0.1.4 (class-utils)',
  '0.1.4 (expand-brackets)',
  '0.1.4 (object-copy)',
  '0.1.4 (snapdragon)',
  '0.1.4 (static-extend)'
]
is-descriptor [
  '0.1.6 (class-utils)',
  '0.1.6 (expand-brackets)',
  '0.1.6 (object-copy)',
  '0.1.6 (snapdragon)',
  '0.1.6 (static-extend)'
]
is-extendable [
  '0.1.1 (expand-brackets)',
  '0.1.1 (extglob)',
  '0.1.1 (findup-sync)',
  '0.1.1 (set-value)',
  '0.1.1 (snapdragon)',
  '0.1.1 (union-value)'
]
is-number [ '3.0.0 (findup-sync)', '3.0.0 (has-values)' ]
kind-of [
  '3.2.2 (class-utils/node_modules/is-accessor-descriptor)',
  '3.2.2 (class-utils/node_modules/is-data-descriptor)',
  '5.1.0 (class-utils)',
  '3.2.2 (expand-brackets/node_modules/is-accessor-descriptor)',
  '3.2.2 (expand-brackets/node_modules/is-data-descriptor)',
  '5.1.0 (expand-brackets)',
  '3.2.2 (findup-sync/node_modules/is-number)',
  '3.2.2 (has-values/node_modules/is-number)',
  '4.0.0 (has-values)',
  '5.1.0 (object-copy/node_modules/is-descriptor)',
  '3.2.2 (object-copy)',
  '3.2.2 (snapdragon-util)',
  '3.2.2 (snapdragon/node_modules/is-accessor-descriptor)',
  '3.2.2 (snapdragon/node_modules/is-data-descriptor)',
  '5.1.0 (snapdragon)',
  '3.2.2 (static-extend/node_modules/is-accessor-descriptor)',
  '3.2.2 (static-extend/node_modules/is-data-descriptor)',
  '5.1.0 (static-extend)',
  '3.2.2 (to-object-path)'
]
locate-path [
  '5.0.0 (@istanbuljs/load-nyc-config)',
  '5.0.0 (nyc)',
  '5.0.0 (pkg-dir)'
]
minimatch [
  '3.1.2 (mocha/node_modules/glob)',
  '5.0.1 (mocha)',
  '5.1.0 (readdir-glob)'
]
minipass [ '3.3.6 (fs-minipass)', '3.3.6 (minizlib)' ]
mkdirp [ '3.0.1 (mocha-junit-reporter)', '1.0.4 (pm2)', '1.0.4 (tar)' ]
ms [
  '2.1.2 (@babel/core)',
  '2.1.2 (@babel/traverse)',
  '2.1.2 (@eslint/eslintrc)',
  '2.1.2 (@humanwhocodes/config-array)',
  '2.1.2 (@koa/router)',
  '2.1.2 (@pm2/agent)',
  '2.1.3 (@pm2/io)',
  '2.1.2 (@pm2/js-api)',
  '2.1.2 (@pm2/pm2-version-check)',
  '2.1.2 (agent-base)',
  '2.1.3 (eslint-import-resolver-node)',
  '2.1.3 (eslint-module-utils)',
  '2.1.2 (eslint)',
  '2.1.2 (get-uri)',
  '2.1.2 (http-proxy-agent)',
  '2.1.2 (https-proxy-agent)',
  '2.1.2 (istanbul-lib-source-maps)',
  '2.1.2 (knex)',
  '2.1.2 (koa)',
  '2.1.2 (mocha-junit-reporter)',
  '2.1.2 (mocha/node_modules/debug)',
  '2.1.3 (mocha)',
  '2.1.3 (needle)',
  '2.1.2 (nock)',
  '2.1.3 (nodemon)',
  '2.1.2 (oidc-provider)',
  '2.1.2 (pac-proxy-agent)',
  '2.1.2 (pm2-axon-rpc)',
  '2.1.2 (pm2-axon)',
  '2.1.2 (pm2-sysmonit)',
  '2.1.2 (pm2)',
  '2.1.2 (proxy-agent)',
  '2.1.3 (require-in-the-middle)',
  '2.1.3 (send)',
  '2.1.2 (socks-proxy-agent)',
  '2.1.2 (superagent)'
]
on-finished [
  '2.4.1 (body-parser)',
  '2.4.1 (express)',
  '2.4.1 (finalhandler)',
  '2.4.1 (send)'
]
p-limit [
  '2.3.0 (@istanbuljs/load-nyc-config)',
  '2.3.0 (nyc)',
  '2.3.0 (pkg-dir)'
]
p-locate [
  '4.1.0 (@istanbuljs/load-nyc-config)',
  '4.1.0 (nyc)',
  '4.1.0 (pkg-dir)'
]
readable-stream [
  '2.3.7 (archiver-utils)',
  '2.3.7 (concat-stream)',
  '2.3.7 (htmlparser2)',
  '2.3.7 (lazystream)',
  '2.3.7 (streamtest)'
]
resolve-from [ '5.0.0 (@istanbuljs/load-nyc-config)', '5.0.0 (nyc)' ]
safe-buffer [
  '5.2.1 (content-disposition)',
  '5.2.1 (express)',
  '5.2.1 (pidusage)',
  '5.2.1 (pm2-sysmonit)',
  '5.2.1 (string_decoder)'
]
semver [
  '6.3.1 (@babel/core)',
  '6.3.1 (@babel/helper-compilation-targets)',
  '7.5.4 (@pm2/agent)',
  '7.5.4 (@pm2/io)',
  '6.3.1 (eslint-config-airbnb-base)',
  '6.3.1 (istanbul-lib-instrument)',
  '6.3.1 (make-dir)'
]
source-map [ '0.8.0-beta.0 (get-stack-trace)', '0.5.7 (snapdragon)' ]
sprintf-js [ '1.1.3 (ip-address)', '1.1.2 (pm2)', '1.1.2 (slonik-sql-tag-raw)' ]
statuses [ '1.5.0 (http-assert)', '1.5.0 (koa)' ]
string_decoder [
  '1.1.1 (archiver-utils)',
  '1.1.1 (concat-stream)',
  '1.1.1 (htmlparser2)',
  '1.1.1 (lazystream)',
  '1.1.1 (streamtest)'
]
supports-color [
  '5.5.0 (@babel/code-frame)',
  '5.5.0 (@babel/highlight)',
  '8.1.1 (mocha)',
  '5.5.0 (nodemon)'
]
tslib [ '1.9.3 (@pm2/io)', '2.6.1 (ast-types)' ]
type-fest [ '0.8.1 (hasha)', '0.12.0 (slonik-sql-tag-raw)' ]
uuid [ '8.3.2 (hyperid)', '8.3.2 (istanbul-lib-processinfo)' ]
```
</details>

It would be better to actively check that express does not pull in its own `qs` module separate to the rest of the project.

Reply:
I've updated the PR:

* more explanation of what's going on
* explicit attempt to load express's `qs` dependency